agent/checks/alias.go

```diff
@@ -0,0 +1,198 @@
+func (c *CheckAlias) Stop() {
+	c.stopLock.Lock()
+	defer c.stopLock.Unlock()
+	if !c.stop {
+		c.stop = true
+		close(c.stopCh)
+	}
+}
+
+// run is invoked in a goroutine until Stop() is called.
+func (c *CheckAlias) run(stopCh chan struct{}) {
+	// If we have a specific node set, then use a blocking query
+	if c.Node != "" {
+		c.runQuery(stopCh)
+		return
+	}
+
+	// Use the local state to match the service.
+	c.runLocal(stopCh)
+}
+
+func (c *CheckAlias) runLocal(stopCh chan struct{}) {
+	// Very important this is buffered as 1 so that we do not lose any
+	// queued updates. This only has to be exactly 1 since the existence
+	// of any update triggers us to load the full health check state.
+	notifyCh := make(chan struct{}, 1)
+	c.Notify.AddAliasCheck(c.CheckID, c.ServiceID, notifyCh)
+	defer c.Notify.RemoveAliasCheck(c.CheckID, c.ServiceID)
+
+	for {
+		select {
+		case <-notifyCh:
+			checks := c.Notify.Checks()
+			checksList := make([]*structs.HealthCheck, 0, len(checks))
+			for _, chk := range checks {
+				checksList = append(checksList, chk)
+			}
+			c.processChecks(checksList)
+
+		case <-stopCh:
+			return
+		}
+	}
+}
+
+func (c *CheckAlias) runQuery(stopCh chan struct{}) {
+	args := c.RPCReq
+	args.Node = c.Node
+	args.AllowStale = true
+	args.MaxQueryTime = 1 * time.Minute
+
+	var attempt uint
+	for {
+		// Check if we're stopped. We fallthrough and block otherwise,
+		// which has a maximum time set above so we'll always check for
+		// stop within a reasonable amount of time.
+		select {
+		case <-stopCh:
+			return
+		default:
+		}
+
+		// Backoff if we have to
+		if attempt > checkAliasBackoffMin {
+			waitTime := (1 << (attempt - checkAliasBackoffMin)) * time.Second
+			if waitTime > checkAliasBackoffMaxWait {
+				waitTime = checkAliasBackoffMaxWait
+			}
+			time.Sleep(waitTime)
+		}
+
+		// Get the current health checks for the specified node.
+		//
+		// NOTE(mitchellh): This currently returns ALL health checks for
+		// a node even though we also have the service ID. This can be
+		// optimized if we introduce a new RPC endpoint to filter both,
+		// but for blocking queries isn't that more efficient since the checks
+		// index is global to the cluster.
+		var out structs.IndexedHealthChecks
+		if err := c.RPC.RPC("Health.NodeChecks", &args, &out); err != nil {
+			attempt++
+			if attempt > 1 {
+				c.Notify.UpdateCheck(c.CheckID, api.HealthCritical,
+					fmt.Sprintf("Failure checking aliased node or service: %s", err))
+			}
+
+			continue
+		}
+
+		attempt = 0 // Reset the attempts so we don't backoff the next
+
+		// Set our index for the next request
+		args.MinQueryIndex = out.Index
+
+		// We want to ensure that we're always blocking on subsequent requests
+		// to avoid hot loops. Index 1 is always safe since the min raft index
+		// is at least 5. Note this shouldn't happen but protecting against this
+		// case is safer than a 100% CPU loop.
+		if args.MinQueryIndex < 1 {
+			args.MinQueryIndex = 1
+		}
```

Comment:
Learned my lesson (cc @banks). Even if this is theoretically not possible, I protect against it.